8329331: Intrinsify Unsafe::setMemory

asgibbons opened the pull request:
This code makes an intrinsic stub for `Unsafe::setMemory` for x86_64.  See [this PR](https://github.com/openjdk/jdk/pull/16760) for discussion around this change.

Overall, making this an intrinsic improves overall performance of `Unsafe::setMemory` by up to 4x for all buffer sizes.

Tested with tier-1 (and full CI).  I've added a table of the before and after numbers for the JMH I ran (`MemorySegmentZeroUnsafe`).

[setMemoryBM.txt](https://github.com/openjdk/jdk/files/14808974/setMemoryBM.txt)

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8329331](https://bugs.openjdk.org/browse/JDK-8329331): Intrinsify Unsafe::setMemory (**Enhancement** - P4)


### Reviewers
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**) ⚠️ Review applies to [7a1d67e5](https://git.openjdk.org/jdk/pull/18555/files/7a1d67e5fbfd10324d27081061fa3da1919a8e1e)
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18555/head:pull/18555` \
`$ git checkout pull/18555`

Update a local copy of the PR: \
`$ git checkout pull/18555` \
`$ git pull https://git.openjdk.org/jdk.git pull/18555/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18555`

View PR using the GUI difftool: \
`$ git pr show -t 18555`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18555.diff">https://git.openjdk.org/jdk/pull/18555.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18555#issuecomment-2027797340)

dholmes-ora reviewed (commented):
This looks like it is still a Draft/work-in-progress. There is only code for x64 and it doesn't appear it will build on other platforms. Also there are still a bunch of `if 0` in the code that should not be there.

dougxc commented:
Wouldn't it be better to do this intrinsification directly in the JIT without calling out to a stub?

asgibbons commented:
> Wouldn't it be better to do this intrinsification directly in the JIT without calling out to a stub?

I believe the code size is too large for a direct JIT intrinsic.  A lot of registers are also used, which may be an issue.

dean-long reviewed (changes requested):
I think the right approach is to turn it into a loop in the IR, which I think is what Doug was implying.  That way C2 can do all its usual optimizations, like unrolling, vectorization, and redundant store elimination (if it is an on-heap primitive array that was just allocated, then there is no need to zero the parts that are being "set").

dholmes-ora commented:
> @dholmes-ora Sorry for the dead code left in. It is gone now. Plus, this was only requested for x86, thus no implementation for other platforms.

Only requested by whom? The JBS issue says nothing about that.

I'm not even sure how this avoids the  `CheckIntrinsics` check for missing intrinsics ... I guess it must only look for some kind of declaration not an actual implementation.

dean-long commented:
As an experiment, couldn't you have the C2 intrinsic redirect to a Java helper that calls putByte() in a loop?

asgibbons commented:
Thanks for all your thoughtful feedback.  I would very much like to take "the right approach"(tm) but I don't have the skill to write IR, especially given that this is an Unsafe block, which is restricted by atomicity and alignment.  I would not know how to prevent the C2 optimizer from vectorizing, or indeed replacing my code with a call to memset().  I'm not sure it would go this far, but in order to remain compliant with the spec I have to prevent it in the future.

This was modeled after the existing implementation of copyMemory, gives good performance (3-5x), and can serve as a template for other platform developers to follow.  They have the expertise for their specific platform(s) which I do not have.

Again, thank you.

dean-long commented:
I went ahead and tried a pure-Java implementation, and it is faster for small sizes (up to 8) and only about 1.5x slower for larger sizes, so that might make for an interesting fallback if there is no customized assembler implementation available or if the size is known to me small.

Ideally, I think we would want C2 to be more aware of setMemory stores, so that it can remove redundant stores, like it does with InitializeNode.

franz1981 commented:
@dean-long in my old PR I have done the same, choosing a configurable cutoff value.

dean-long commented:
Thanks, I see that my ideas have pretty much already been discussed in https://github.com/openjdk/jdk/pull/16760.  I might have missed it, but has the possibility of always setting the aligned interior region with 8 byte stores been discussed?  A literal reading of the javadoc seems to disallow it, but it seems like it should be allowed based on memory coherence.  Only the unaligned head and tail would need special treatment.

asgibbons commented:
As an experiment I added the java code that @franz1981 supplied and ran performance vs. the intrinsic stub.  I used 128 bytes as the cutoff value as in that code.  I saw about 0.75 to 1ns improvement for sizes of 1 or 2 bytes only.  Anything larger and the stub performed better.

@mcimadamore Is there any way to disable some of the optimizations C2 will attempt on the IR?  We need to maintain atomicity, so vectorization shouldn't occur, for instance.  This seems like a rat-hole that would need constant maintenance as C2 optimizations get better.

mcimadamore commented:
> Is there any way to disable some of the optimizations C2 will attempt on the IR? We need to maintain atomicity, so vectorization shouldn't occur, for instance. This seems like a rat-hole that would need constant maintenance as C2 optimizations get better.

Sorry, I do not know that (I'm not a C2 engineer :-) ). One small observation: how important is atomicity in the "full off-heap case" ? E.g. if a `setMemory` is occurring at a location that is provably off-heap (and we should have ways to detect that, we do that also for other unsafe memory access routines), then perhaps the atomicity requirement can go (as I suppose that requirement is there for the Java Memory Model) ?

So, perhaps, while we might not be able to fully optimize for on-heap access, we might be able to do so for off-heap access (which is an important case for FFM).

sviswa7 commented:
To me doing the compiler intrinsic for Unsafe setMemory looked to be the right first step as in this PR. There is precedence with Unsafe copyMemory intrinsic so we are in sync with what is done before. We could request Fei Yang/Hamlin Li for RISC-V intrinsic and Bhavana Kilambi/Nick Gasson for AARCH64 intrinsic as follow up PRs.

sviswa7 commented on src/hotspot/share/opto/library_call.cpp:
```diff
@@ -4945,6 +4946,57 @@ bool LibraryCallKit::inline_unsafe_copyMemory() {
 }
 
+bool LibraryCallKit::inline_unsafe_setMemory() {
+  if (callee()->is_static())  return false;  // caller must have the capability!
```
Also need to return false if StubRoutines::unsafe_setmemory() == nullptr.

asgibbons replied:
Will do.

sviswa7 commented on src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:
```diff
@@ -2476,6 +2481,281 @@ address StubGenerator::generate_unsafe_copy(const char *name,
+// Examines the alignment of the operands and dispatches
+// to an int, short, or byte fill loop.
+//
+address StubGenerator::generate_unsafe_setmemory(const char *name,
+                                                 address byte_fill_entry) {
```
Need to add UnsafeSetMemoryMark on similar lines as UnsafeCopyMemoryMark to handle page error.

asgibbons replied:
Will do.

sviswa7 commented on src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:
```diff
@@ -2476,6 +2481,281 @@ address StubGenerator::generate_unsafe_copy(const char *name,
+    Label L_exit, L_fillQuadwords, L_fillDwords, L_fillBytes;
+
+    setup_arg_regs(3);
+#undef dest
+#define dest rdi
+#undef size
+#define size rsi
+#undef wide_value
+#define wide_value rax
+#undef rScratch1
+#define rScratch1 rcx
+#undef byteVal
+#define byteVal rdx
+#undef rScratch3
+#define rScratch3 r8
+#undef rScratch4
+#define rScratch4 r11
```
We could do this setup using const Register declaration instead of using #undef/#define pair.

asgibbons replied:
We discussed this and the #define option was your preferred method since the registers are being re-used.  Do you want this changed back again?

sviswa7 commented on src/hotspot/cpu/x86/macroAssembler_x86.cpp:
```diff
@@ -5985,13 +5985,13 @@ void MacroAssembler::generate_fill(BasicType t, bool aligned,
     movw(Address(to, 0), value);
     addptr(to, 2);
-    subl(count, 1<<(shift-1));
+    subptr(count, 1<<(shift-1));
```
At line 5968 also we need the change from cmpl to cmpptr.
cmpl(count, 2<<shift); // Short arrays (< 8 bytes) fill by element

asgibbons replied:
Will do.

sviswa7 commented on src/hotspot/cpu/x86/macroAssembler_x86.cpp:
```diff
@@ -6047,14 +6047,14 @@ void MacroAssembler::generate_fill(BasicType t, bool aligned,
           vpbroadcastd(xtmp, xtmp, Assembler::AVX_512bit);
 
-          subl(count, 16 << shift);
+          subptr(count, 16 << shift);
```
At line 6045 also the cmpl should change to cmpptr:
cmpl(count, VM_Version::avx3_threshold());

asgibbons replied:
Will do.

asgibbons commented:
I added the framework for setting unsafe access marks within all platforms, and fixed a bug with the Linux 32-bit runtime tests.  Adding stub intrinsic for setMemory0 for other platforms should be easier now.

Passes CI testing.

sviswa7 commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -8336,6 +8336,11 @@ class StubGenerator: public StubCodeGenerator {
       UnsafeCopyMemory::create_table(8);
     }
 
+    // Initialize table for fill memory check.
+    if (UnsafeCopyMemory::_table == nullptr) {
+      UnsafeCopyMemory::create_table(8);
+    }
+
```
Did you mean to initialize UnsafeSetMemory::_table here instead?

asgibbons replied:
Yes.  Good catch.

sviswa7 commented on src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:
```diff
@@ -152,6 +152,11 @@ void StubGenerator::generate_arraycopy_stubs() {
   StubRoutines::_arrayof_jshort_fill = generate_fill(T_SHORT, true, "arrayof_jshort_fill");
   StubRoutines::_arrayof_jint_fill = generate_fill(T_INT, true, "arrayof_jint_fill");
 
+// #ifdef _LP64
```
We could remove the #ifdef _LP64, #endif commented pair.

asgibbons replied:
Done.

sviswa7 commented on src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:
```diff
@@ -727,6 +732,7 @@ address StubGenerator::generate_disjoint_copy_avx3_masked(address* entry, const
   __ ret(0);
 
   if (MaxVectorSize == 64) {
+    UnsafeCopyMemoryMark ucmm(this, !is_oop && !aligned, false, ucme_exit_pc);
```
This is not related to Unsafe::setMemory?

asgibbons replied:
No.  Reviewing the code I saw this as a potential error, as `arraycopy_avx3_large` could cause a SIGBUS which wouldn't be caught.  It conforms to the other instances of copy in the code.  I think it was missed by the original developer.

sviswa7 commented on src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:
```diff
@@ -2476,6 +2480,287 @@ address StubGenerator::generate_unsafe_copy(const char *name,
+      {
+        UnsafeSetMemoryMark usmm(this, true, true);
+
+        __ generate_fill(T_BYTE, false, c_rarg0, c_rarg1, r11, rax, xmm0);
```
We will be duplicating the code gen for generate_fill here? Could we not do a tail call to _jbyte_fill here and add UnsafeSetMemoryMark inside _jbyte_fill?

asgibbons replied:
It would not be appropriate to add set memory marks to the existing _jbyte_fill as it is being used by other routines, and the effect of the mark will be very hard to track down (if any).

Are you *sure* we want to do that?

sviswa7 replied:
Yes we want to do that. It is all guarded by thread->doing_unsafe_access() which is only true when we are getting to this code from unsafe. Similar technique is used in copyMemory as well.

asgibbons replied:
Done.

asgibbons commented on src/hotspot/share/opto/library_call.cpp:
```diff
@@ -4948,6 +4949,57 @@ bool LibraryCallKit::inline_unsafe_copyMemory() {
   return true;
 }
 
+bool LibraryCallKit::inline_unsafe_setMemory() {
```
Done

asgibbons commented on src/hotspot/share/opto/runtime.cpp:
```diff
@@ -771,6 +771,26 @@ const TypeFunc* OptoRuntime::void_void_Type() {
+  fields[argp++] = TypePtr::NOTNULL;    // dest
+  fields[argp++] = TypeLong::LONG;      // size
+  fields[argp++] = Type::HALF;          // size
+  fields[argp++] = TypeInt::INT;        // bytevalue
```
Should be TypeInt::UBYTE.  Changed.

asgibbons commented on src/hotspot/share/runtime/sharedRuntime.cpp:
```diff
@@ -177,6 +177,8 @@ uint SharedRuntime::_slow_array_copy_ctr=0;
 uint SharedRuntime::_find_handler_ctr=0;
 uint SharedRuntime::_rethrow_ctr=0;
 
+uint SharedRuntime::_unsafe_set_memory_ctr=0;
+
```
Done

asgibbons commented on src/hotspot/share/runtime/sharedRuntime.cpp:
```diff
@@ -1989,6 +1991,8 @@ void SharedRuntime::print_statistics() {
   if (_find_handler_ctr) tty->print_cr("%5u find exception handler", _find_handler_ctr);
   if (_rethrow_ctr) tty->print_cr("%5u rethrow handler", _rethrow_ctr);
 
+  if (_unsafe_set_memory_ctr) tty->print_cr("%5u unsafe set memorys", _unsafe_set_memory_ctr);
```
Done.

asgibbons commented on src/hotspot/share/runtime/sharedRuntime.hpp:
```diff
@@ -542,6 +542,8 @@ class SharedRuntime: AllStatic {
   static uint _generic_array_copy_ctr;     // Slow-path includes type decoding
   static uint _slow_array_copy_ctr;        // Slow-path failed out to a method call
 
+  static uint _unsafe_set_memory_ctr;      // Slow-path includes alignment checks
+
```
Done.

asgibbons commented on test/jdk/sun/misc/CopyMemory.java:
```diff
@@ -207,6 +207,11 @@ private static Unsafe getUnsafe() throws Exception {
     public static void main(String[] args) throws Exception {
         Unsafe unsafe = getUnsafe();
 
+        // long seed = 6742745864802755133L;
+        long seed = random.nextLong();
+        random.setSeed(seed);
+        System.out.println("Seed set to "+ seed);
+
```
Yes, but I believe we should adopt this for the future since reproducing random test failures is extremely difficult without knowing the seed of the RNG.

asgibbons replied:
Removed.

sviswa7 commented on src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:
```diff
@@ -2476,6 +2478,286 @@ address StubGenerator::generate_unsafe_copy(const char *name,
+    {
+      const Register byteVal = rdx;
+
+      // Propagate byte to full Register
```
The comment refers to lines 2524-2526, please move it down.

sviswa7 replied:
Still continuing to look through StubGenerator::generate_unsafe_setmemory(), more comments to come. Thank you for your patience.

asgibbons replied:
Comment moved.

sviswa7 commented on src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:
```diff
@@ -2476,6 +2478,286 @@ address StubGenerator::generate_unsafe_copy(const char *name,
+  {
+    Label L_exit, L_fillQuadwords, L_fillDwords, L_fillBytes;
+
+    setup_arg_regs(3);
```
A comment stating the placement of dest, size, and byteVal after call to setup_arg_regs() would be very helpful.

asgibbons replied:
Done.

sviswa7 commented on src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:
```diff
@@ -2477,6 +2483,220 @@ address StubGenerator::generate_unsafe_copy(const char *name,
+    __ BIND(L_fillBytes);
+    {
+      const Register byteVal = rdx;
+#ifdef _WIN32
+      __ movq(rcx, rdi); // Restore c_rarg*
+      __ movq(r8, rdx);
+      __ movq(rdx, rsi);
+      restore_arg_regs();
+#endif
```
This is stubGenerator_x86_64.cpp 64bit specific, so WIN32 portion could be removed?

JornVernee replied:
`_WIN32` is also defined for 64 bit Windows

sviswa7 replied:
Thanks, I missed that.

asgibbons replied:
I'm changing the scheme here to c_rargX, so no need for any windows-specific stuff.  I added it when I needed 4 scratch registers.

sviswa7 commented on src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:
```diff
@@ -2477,6 +2483,220 @@ address StubGenerator::generate_unsafe_copy(const char *name,
+
+    __ BIND(L_fillBytes);
+    {
+      const Register byteVal = rdx;
```
This could be removed.

sviswa7 commented on src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:
```diff
@@ -2477,6 +2483,220 @@ address StubGenerator::generate_unsafe_copy(const char *name,
+                                                 address unsafe_byte_fill) {
+  __ align(CodeEntryAlignment);
+  StubCodeMark mark(this, "StubRoutines", name);
+  address start = __ pc();
+
```
We are missing the __ enter() here?

sviswa7 reviewed (commented):
These are my last set of comments. Rest looks good to me.

sviswa7 commented:
/reviewers 2

sviswa7 reviewed (approved):
PR looks good to me now. Thanks a lot for considering all the inputs.

asgibbons commented:
I ran the benchmark again using a branch from openjdk today against this modification.  Performance increased an average of 4.98x (max - 7.52x (unsafe 64-byte aligned), min - 3.02x (panama 255-byte aligned), stddev - 0.94x).

I'll integrate after a second review has been done.  Thanks!

sviswa7 commented:
@vnkozlov Could you please review this PR from @asgibbons? Looking forward to your inputs.

jatin-bhateja reviewed (commented):
Hi @asgibbons Please add a new test / extend an existing test for SIGBUS violation testing test/hotspot/jtreg/runtime/Unsafe/InternalErrorTest.java

asgibbons commented on test/micro/org/openjdk/bench/java/lang/foreign/MemorySegmentZeroUnsafe.java:
```diff
@@ -0,0 +1,70 @@
+package org.openjdk.bench.java.lang.foreign;
```
Added

asgibbons commented on src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:
```diff
@@ -2477,6 +2483,201 @@ address StubGenerator::generate_unsafe_copy(const char *name,
+
+    // Propagate byte to full Register
+    __ movzbl(rScratch1, byteVal);
+    __ mov64(wide_value, 0x0101010101010101);
```
Fixed.

JornVernee commented on src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:
```diff
@@ -2477,6 +2483,201 @@ address StubGenerator::generate_unsafe_copy(const char *name,
+  // At this point, we know the lower bits of size are zero
+  __ shrq(size, shiftval);
+  // size now has number of X-byte chunks (2, 4 or 8)
+  __ cmpq(size, 8);
+  __ jccb(Assembler::below, L_Tail);
+
+  // Number of (8*X)-byte chunks into rScratch1
+  __ movq(tmp, size);
+  __ shrq(tmp, 3);
```
`shr` [sets the zero flag][1], so I think you can just move the jump to after the shift and avoid a separate comparison?

```suggestion  
  // Number of (8*X)-byte chunks into rScratch1
  __ movq(tmp, size);
  __ shrq(tmp, 3);
  __ jccb(Assembler::zero, L_Tail);
```

[1]: https://www.felixcloutier.com/x86/sal:sar:shl:shr#flags-affected

asgibbons replied:
Good catch.  Done.

vnkozlov commented on src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:
```diff
@@ -4009,8 +4009,8 @@ void StubGenerator::generate_initial_stubs() {
-  if (UnsafeCopyMemory::_table == nullptr) {
-    UnsafeCopyMemory::create_table(16);
+  if (UnsafeMemoryAccess::_table == nullptr) {
+    UnsafeMemoryAccess::create_table(26);
```
This needs comment

asgibbons replied:
I added 10 to the table size because I knew I was going to add 7 places where a mark was required for setMemory.  I left 3 for safety.  The algorithm changed so only 4 are needed.

The algorithm has since changed, so I changed this to:

`UnsafeCopyMemory::create_table(16 + 4); // 16 for copyMemory; 4 for setMemory`

I did a similar change to all other table creation numbers.

asgibbons replied:
I took what was there and added 10, since the original algorithm required 7 marks for setMemory. I added 3 for safety. The algorithm changed so only 4 are needed, so I changed this to:

`UnsafeCopyMemory::create_table(16 + 4); // 16 for copyMemory; 4 for setMemory`

Hope this is clearer.

vnkozlov commented on src/hotspot/share/utilities/copy.hpp:
```diff
@@ -300,7 +300,6 @@ class Copy : AllStatic {
  protected:
   inline static void shared_disjoint_words_atomic(const HeapWord* from,
                                                   HeapWord* to, size_t count) {
-
```
I don't think this justify to change the file.

asgibbons replied:
Reverted.

asgibbons replied:
Un-done.  Thanks.

vnkozlov commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -8332,8 +8332,8 @@ class StubGenerator: public StubCodeGenerator {
     // Initialize table for copy memory (arraycopy) check.
-    if (UnsafeCopyMemory::_table == nullptr) {
-      UnsafeCopyMemory::create_table(8);
+    if (UnsafeMemoryAccess::_table == nullptr) {
+      UnsafeMemoryAccess::create_table(18);
```
Needs comment explaining 18 number

asgibbons replied:
Hmmm...  There was no comment explaining the 8 number :-).  I added 10 to the table size because I knew I was going to add 7 places where a mark was required.  I left 3 for safety.

The algorithm has since changed, so I changed this to:

`UnsafeCopyMemory::create_table(8 + 4); // 8 for copyMemory; 4 for setMemory`

I did a similar change to all other table creation numbers.

asgibbons replied:
Well, the 8 number wasn't explained either :-).  I took what was there and added 10, since the original algorithm required 7 marks for setMemory.  I added 3 for safety.  The algorithm changed so only 4 are needed, so I changed this to:

`UnsafeCopyMemory::create_table(8 + 4); // 8 for copyMemory; 4 for setMemory`

Hope this is clearer.

vnkozlov reviewed (changes requested):
General comment/suggestion before I dive into review.
Can we do renaming `UnsafeCopyMemory*` -> `UnsafeMemory*` in follow up RFE. This change hides the real change.

asgibbons commented on src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:
```diff
@@ -2477,6 +2483,201 @@ address StubGenerator::generate_unsafe_copy(const char *name,
+    const Register byteVal = c_rarg2;
+    const Register wide_value = rax;
+    const Register rScratch1 = r10;
+
```
Done.

asgibbons commented on src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:
```diff
@@ -2477,6 +2483,201 @@ address StubGenerator::generate_unsafe_copy(const char *name,
+    // Set up for tail call to previously generated byte fill routine
+    // Parameter order is (ptr, byteVal, size)
+    __ xchgq(c_rarg1, c_rarg2);
+    __ pop(rbp);    // Clear effect of enter()
```
No special reason.  I've changed it since it seems to provide more clarity.

asgibbons commented on src/hotspot/share/opto/library_call.cpp:
```diff
@@ -4948,6 +4949,59 @@ bool LibraryCallKit::inline_unsafe_copyMemory() {
+  null_check_receiver();  // null-check receiver
+  if (stopped())  return true;
+
+  if (StubRoutines::unsafe_setmemory() == nullptr) return false;
```
Removed.

asgibbons replied:
Removed.

asgibbons commented on src/hotspot/share/opto/runtime.cpp:
```diff
@@ -771,6 +771,30 @@ const TypeFunc* OptoRuntime::void_void_Type() {
+// uchar byte
+const TypeFunc* OptoRuntime::make_setmemory_Type() {
+  // create input type (domain)
+  int num_args      = 4;
```
It is.  It is there due to copy/paste from the other 10 places that also have the same redundant variable declaration.  I've removed it from here, but I think I'll be asked to submit a separate PR if I remove it from the other locations.

Note that it's also redundant in `make_arraycopy_Type()`.

asgibbons replied:
This is a copy/paste.  There are 10 places in this file that declare `num_args` and only use it to assign to `argcnt`.  I've changed this one, but I think I'll be asked to submit a separate PR to change the other 9.

asgibbons commented on src/hotspot/share/opto/runtime.cpp:
```diff
@@ -771,6 +771,30 @@ const TypeFunc* OptoRuntime::void_void_Type() {
+  const Type** fields = TypeTuple::fields(argcnt);
+  int argp = TypeFunc::Parms;
+  fields[argp++] = TypePtr::NOTNULL;    // dest
+  fields[argp++] = TypeLong::LONG;      // size
+  fields[argp++] = Type::HALF;          // size
```
I don't understand this well enough to be confident in the change.  Can you please verify that I've changed it properly?

asgibbons replied:
I'm not confident in my change for this.  Could you please take another look at what I've done?

JornVernee replied:
Your latest version looks good to me.

asgibbons commented:
@vnkozlov I un-did the name change and will submit a separate request for re-naming.  Thanks.

vnkozlov commented on src/hotspot/share/opto/runtime.cpp:
```diff
@@ -771,6 +771,29 @@ const TypeFunc* OptoRuntime::void_void_Type() {
+  int argp = TypeFunc::Parms;
+  fields[argp++] = TypePtr::NOTNULL;        // dest
+  fields[argp++] = TypeX_X;                 // size
+  LP64_ONLY(fields[argp++] = Type::HALF);  // size
```
Nit: align `/`

asgibbons replied:
Done

vnkozlov commented on src/hotspot/share/utilities/copy.hpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2003, 2022, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2003, 2024, Oracle and/or its affiliates. All rights reserved.
```
I forgot to undo year change in this file.

asgibbons replied:
Yup. Done.

vnkozlov reviewed (commented):
This looks good. I only have question about long vs short jumps in stub's code.

asgibbons commented on src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:
```diff
@@ -2477,6 +2483,202 @@ address StubGenerator::generate_unsafe_copy(const char *name,
+      do_setmemory_atomic_loop(USM_SHORT, dest, size, wide_value, rScratch1,
+                               L_exit, _masm);
+    }
+    __ jmp(L_exit);
```
The old code required a long jump due to the size of `do_setmemory_atomic_loop` but has since been refactored.  The `jmp(Label)` code will generate a short jump provided the label has been defined and is in range.  Otherwise a long jump is generated.

Changed to `jmpb`

vnkozlov commented:
Good. I will submit our testing.

vnkozlov commented:
`runtime/Unsafe/InternalErrorTest.java` test SIGBUS when run with `-Xcomp` (and other flags in test's @run command):
```
#  SIGBUS (0xa) at pc=0x0000000119514760, pid=63021, tid=28163
#
# JRE version: Java(TM) SE Runtime Environment (23.0) (fastdebug build 23-internal-2024-04-19-2326152.vladimir.kozlov.jdkgit2)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (fastdebug 23-internal-2024-04-19-2326152.vladimir.kozlov.jdkgit2, compiled mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-amd64)
# Problematic frame:
# v  ~StubRoutines::jbyte_fill 0x0000000119514760
```

asgibbons commented on src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:
```diff
@@ -2477,6 +2483,202 @@ address StubGenerator::generate_unsafe_copy(const char *name,
+  for (int i = 0; i < 8; i++) {
+    switch (type) {
+      case USM_SHORT:
+        __ movw(Address(dest, (2 * i)), wide_value);
```
My understanding is the spec requires the appropriate-sized write based on alignment and size.  This is why there's no 128-bit or 256-bit store loops.

asgibbons commented on src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:
```diff
@@ -2477,6 +2483,202 @@ address StubGenerator::generate_unsafe_copy(const char *name,
+  __ BIND(L_Loop);
+
+  // Unroll 8 stores
+  for (int i = 0; i < 8; i++) {
+    switch (type) {
+      case USM_SHORT:
+        __ movw(Address(dest, (2 * i)), wide_value);
+        break;
+      case USM_DWORD:
+        __ movl(Address(dest, (4 * i)), wide_value);
+        break;
+      case USM_QUADWORD:
+        __ movq(Address(dest, (8 * i)), wide_value);
+        break;
+    }
+  }
```
I believe the spec is specific on the size of the store required given alignment and size.  I want to honor that spec even though wider stores could be done in many cases.

asgibbons commented:
The SIGBUS was improper scoping of the UnsafeCopyMemoryMark.  The change is:

`   {`
`     // Add set memory mark to protect against unsafe accesses faulting`
`-    UnsafeCopyMemoryMark(this, ((t == T_BYTE) && !aligned), true);`
`+    UnsafeCopyMemoryMark usmm(this, ((t == T_BYTE) && !aligned), true);`
`     __ generate_fill(t, aligned, to, value, r11, rax, xmm0);`
`   }`

asgibbons commented:
@vnkozlov Thanks for the feedback.  Can you please start the testing again?  I'd appreciate it.

vnkozlov commented:
Before I do testing, please sync with mainline.

asgibbons commented:
Merge done.

vnkozlov reviewed (approved):
My testing passed. Good.

asgibbons commented:
/integrate

Thank you all for the reviews.

jatin-bhateja commented:
/sponsor

TobiHartmann commented:
This introduced a regression, see [JDK-8331033](https://bugs.openjdk.org/browse/JDK-8331033).
